### ISSUE TITLE
Segfault in minimized parallel external source example

### ISSUE THREAD

**Issue by tk-woven** [NVIDIA/DALI#5131]
### Describe the question.

# Introduction

I ran into a segfault when integrating DALI in another repo. I have reduced the official example provided in the [parallel external source docs](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/data_loading/parallel_external_source.html) to an MRE that resembles my situation.

# Environment

Ubuntu 20.04, Python 3.8, CUDA 12.0 (Driver Version: 525.125.06).

I have reproduced the MRE on:

* nvidia-dali-cuda120
* nvidia-dali-cuda110-1.30.0
* nvidia-dali-nightly-cuda110-1.32.0.dev20231024

**requirements.txt** (for CUDA 11-based installation):

```
numpy==1.24.4
pytest==7.4.3

--extra-index-url https://developer.download.nvidia.com/compute/redist
nvidia-dali-cuda110
```

# MRE

**repro.py**
```py
import numpy as np
from nvidia.dali import fn, pipeline_def, types          
                                                         
DATASET_SIZE = 21                                         
BATCH_SIZE = 5   
BATCHES_PER_EPOCH = DATASET_SIZE // BATCH_SIZE
                                                 
class ExternalInputCallable:                
    def __call__(self, sample_info):        
        if sample_info.iteration >= BATCHES_PER_EPOCH:
            raise StopIteration()                          
        return (np.ones((480, 640, 3), dtype=np.uint8),)     
                                                      
@pipeline_def(                                          
    batch_size=BATCH_SIZE,         
    num_threads=2,                 
    device_id=0,                      
    py_num_workers=4,                      
    py_start_method="spawn",                
)                                   
def parallel_pipeline():            
    [jpegs] = fn.external_source(       
        source=ExternalInputCallable(), 
        num_outputs=1,                                     
        batch=False,                                   
        parallel=True,                                
        dtype=[types.UINT8]                           
    )                                  
    jpegs = fn.resize(jpegs, size=(256, 512))
    return jpegs

def test_main():
    parallel_pipe = parallel_pipeline()
    parallel_pipe.build()
    parallel_pipe.run()
                                                   
if __name__ == "__main__":
    test_main()
```

Repro:

```
$ python3 repro.py
Segmentation fault (core dumped)
```

The tail end of a GDB session had this to say (debugging the CUDA 11-based DALI installation) (reformatted line width):

```
Thread 1 "python3" received signal SIGSEGV, Segmentation fault.
0x00007fffb6548381 in dali::mm::binning_resource_base<
  dali::cuda_for_dali::memory_kind::host,
  -1,
  dali::cuda_for_dali::memory_resource<dali::cuda_for_dali::memory_kind::host>
>::do_deallocate(void*, unsigned long, unsigned long) ()
   from tk-woven/.venv/lib/python3.8/site-packages/nvidia/dali/libdali_core.so
```

## Quick Fix

If I inline the body of `test_main` into `if __name__ == "__main__":`, I cannot reproduce the segfault. Coincidentally, if I set `py_num_workers=0` in the pipeline definition in the MRE, this also causes a segfault even with `test_main` inlined under the `__main__` check.

# Question

Am I missing something obvious with the scoping of the calls? Ultimately, I have some DALI-based tests I am trying to run with pytest, and if we were to invoke `pytest repro.py`, we will currently encounter the segfault given the MRE above.

### Check for duplicates

- [X] I have searched the [open bugs/issues](https://github.com/NVIDIA/DALI/issues) and have found no duplicates for this bug report

**Comment by JanuszL** [comment]
Hi @tk-woven,

Thank you for reaching out. I can confirm that this problem reproduces on my end as well.
Your reproduction script is very good and helpful. I appreciate your time and effort to provide it.
Let us investigate the issue and get back to you soon.

**Comment by tk-woven** [comment]
Hi @JanuszL , thank you kindly for looking into this. I'll stand by.

**Comment by mzient** [comment]
@tk-woven Fix pending, #5133 - it should be available in the next successful nightly build.

**Comment by tk-woven** [comment]
Thank you @mzient . I'll give the next successful nightly build a shot.

**Comment by tk-woven** [comment]
Hi @mzient , is there a place where I can easily track the latest available nightly build and which commits it contains by chance? Currently I just use pip to see if another build is available, once a day. I just installed `nvidia-dali-nightly-cuda110-1.32.0.dev20231030` but would like to confirm whether that build is supposed to have the fix. Thank you kindly.

**Comment by JanuszL** [comment]
Hi @tk-woven,

You can browse [this](https://developer.download.nvidia.com/compute/redist/nightly/nvidia-dali-nightly-cuda110/) to check new builds.
Regarding the commit - just issue:
```pip show nvidia-dali-nightly-cuda110```

**Comment by tk-woven** [comment]
@JanuszL , thank you kindly.

All, I have confirmed the fix on my end via nightly `1.32.0.dev20231101`. Thank you very much for your prompt fix!
